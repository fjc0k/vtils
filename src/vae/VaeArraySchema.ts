import { Nullable } from '../types/index.ts'
import { isArray } from '../utils/index.ts'
import { VaeLocale, VaeLocaleMessage } from './VaeLocale.ts'
import { VaeSchema, VaeSchemaOf } from './VaeSchema.ts'

export type VaeArraySchemaElementOf<T> = T extends Array<infer X>
  ? VaeSchemaOf<X>
  : never

export class VaeArraySchema<
  T0 extends Nullable<any[]> = any[],
  T extends NonNullable<T0> = NonNullable<T0>,
> extends VaeSchema<T0> {
  constructor(
    element?: VaeArraySchemaElementOf<T>,
    message: VaeLocaleMessage = VaeLocale.array.type,
  ) {
    super({
      type: 'array',
    })

    this.check({
      fn: isArray,
      message: message,
    })

    if (element) {
      this.element(element)
    }
  }

  element(element: VaeArraySchemaElementOf<T>) {
    return this.check({
      fn: element,
      message: '',
      tag: 'element',
    })
  }

  nonempty(message: VaeLocaleMessage = VaeLocale.array.nonempty) {
    return this.check({
      fn: v => v.length > 0,
      message: message,
      tag: 'nonempty',
    })
  }

  min(value: number, message: VaeLocaleMessage = VaeLocale.array.min) {
    return this.check({
      fn: v => v.length >= value,
      message: message,
      messageParams: {
        min: value,
      },
      tag: 'min',
    })
  }

  max(value: number, message: VaeLocaleMessage = VaeLocale.array.max) {
    return this.check({
      fn: v => v.length <= value,
      message: message,
      messageParams: {
        max: value,
      },
      tag: 'max',
    })
  }

  length(value: number, message: VaeLocaleMessage = VaeLocale.array.length) {
    return this.check({
      fn: v => v.length === value,
      message: message,
      messageParams: {
        length: value,
      },
      tag: 'length',
    })
  }
}
