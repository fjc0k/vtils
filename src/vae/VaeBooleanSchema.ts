import { Nullable } from '../types'
import { isBoolean } from '../utils'
import { VaeLocale, VaeLocaleMessage } from './VaeLocale'
import { VaeSchema } from './VaeSchema'

export class VaeBooleanSchema<
  T0 extends Nullable<boolean> = boolean,
> extends VaeSchema<T0> {
  constructor(message: VaeLocaleMessage = VaeLocale.boolean.type) {
    super({
      type: 'boolean',
    })
    this.transform(Boolean as any).check({
      fn: isBoolean,
      message: message,
    })
  }

  true(message: VaeLocaleMessage = VaeLocale.boolean.true) {
    return this.check({
      fn: v => v === true,
      message: message,
      tag: 'true',
    })
  }

  false(message: VaeLocaleMessage = VaeLocale.boolean.false) {
    return this.check({
      fn: v => v === false,
      message: message,
      tag: 'false',
    })
  }
}
