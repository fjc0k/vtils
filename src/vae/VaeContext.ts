import { VaeIssue } from './VaeIssue'
import { VaeSchemaPath } from './VaeSchema'

export class VaeContext {
  oldPath: VaeSchemaPath = []

  path: VaeSchemaPath = []

  issues: VaeIssue[] = []

  setPath = (path: VaeSchemaPath) => {
    this.oldPath = this.path.slice()
    this.path = path
  }

  restorePath = () => {
    this.path = this.oldPath
  }

  addIssue(issue: VaeIssue) {
    this.issues.push(issue)
  }
}
