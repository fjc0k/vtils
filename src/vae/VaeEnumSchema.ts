import { includes, isArray, values } from '../utils'
import { VaeLocale, VaeLocaleMessage } from './VaeLocale'
import { VaeSchema } from './VaeSchema'

export class VaeEnumSchema<T extends any = any> extends VaeSchema<T> {
  constructor(
    value: T[] | Record<any, T>,
    message: VaeLocaleMessage = VaeLocale.enum.type,
  ) {
    super({
      type: 'enum',
    })

    const enumValues = isArray(value) ? value : values(value)
    this.check({
      fn: v => includes(enumValues, v),
      message: message,
      messageParams: {
        enum: enumValues,
      },
    })
  }
}
