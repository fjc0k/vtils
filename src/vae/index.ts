/**
 * vae 数据验证。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

export * from './vae'
