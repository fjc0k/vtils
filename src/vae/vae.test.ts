import { VaeSchemaOf, v } from './vae'

describe('vae', () => {
  test('string', () => {
    expect(
      v
        .string()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(v.string().parse('')).toMatchSnapshot()
    expect(v.string().parse('hello')).toMatchSnapshot()

    expect(
      v
        .string()
        .required()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(
      v
        .string()
        .default('hello')
        .required()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(v.string().required().parse('')).toMatchSnapshot()
    expect(v.string().required().emptyable().parse('')).toMatchSnapshot()
    expect(v.string().required().parse('hello')).toMatchSnapshot()

    expect(v.string().required().min(6).parse('hello')).toMatchSnapshot()
    expect(v.string().required().max(2).parse('hello')).toMatchSnapshot()
    expect(v.string().required().length(3).parse('hello')).toMatchSnapshot()

    expect(v.string().required().email().parse('hello')).toMatchSnapshot()
    expect(
      v.string().required().email().parse('hello@gmail.com'),
    ).toMatchSnapshot()

    expect(
      v.string().required().url().parse('hello@gmail.com'),
    ).toMatchSnapshot()
    expect(
      v.string().required().url().parse('https://baidu.com'),
    ).toMatchSnapshot()

    expect(v.string().required().regex(/^\d+$/).parse('k2')).toMatchSnapshot()
    expect(v.string().required().regex(/^\d+$/).parse('234')).toMatchSnapshot()

    expect(v.string().required().includes('dd').parse('k2')).toMatchSnapshot()
    expect(
      v.string().required().includes('dd').parse('23dd4'),
    ).toMatchSnapshot()

    expect(v.string().required().startsWith('dd').parse('k2')).toMatchSnapshot()
    expect(
      v.string().required().startsWith('dd').parse('dd234'),
    ).toMatchSnapshot()

    expect(v.string().required().endsWith('dd').parse('k2')).toMatchSnapshot()
    expect(
      v.string().required().endsWith('dd').parse('234dd'),
    ).toMatchSnapshot()

    expect(v.string().required().phoneNumber().parse('k2')).toMatchSnapshot()
    expect(
      v.string().required().phoneNumber().parse('18080000000'),
    ).toMatchSnapshot()

    expect(v.string().required().idCardNumber().parse('k2')).toMatchSnapshot()
    expect(
      v.string().required().idCardNumber().parse('110101202305033210'),
    ).toMatchSnapshot()

    expect(
      v
        .string()
        .required()
        .trim()
        .idCardNumber()
        .parse('   110101202305033210'),
    ).toMatchSnapshot()
  })

  test('number', () => {
    expect(
      v
        .number()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(
      v.number().parse(
        // @ts-expect-error
        '',
      ),
    ).toMatchSnapshot()
    expect(
      v.number().parse(
        // @ts-expect-error
        '3',
      ),
    ).toMatchSnapshot()
    expect(v.number().parse(3)).toMatchSnapshot()

    expect(
      v
        .number()
        .required()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(
      v
        .number()
        .default(3)
        .required()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(
      v.number().required().parse(
        // @ts-expect-error
        '',
      ),
    ).toMatchSnapshot()
    expect(v.number().required().parse(3)).toMatchSnapshot()

    expect(v.number().required().min(6).parse(0)).toMatchSnapshot()
    expect(v.number().required().max(2).parse(0)).toMatchSnapshot()
    expect(v.number().required().lessThan(2).parse(0)).toMatchSnapshot()
    expect(v.number().required().moreThan(2).parse(0)).toMatchSnapshot()

    expect(v.number().required().integer().parse(0)).toMatchSnapshot()
    expect(v.number().required().integer().parse(-1)).toMatchSnapshot()
    expect(v.number().required().integer().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().integer().parse(1)).toMatchSnapshot()

    expect(v.number().required().positive().parse(1)).toMatchSnapshot()
    expect(v.number().required().positive().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().positive().parse(-1)).toMatchSnapshot()
    expect(v.number().required().positive().parse(0)).toMatchSnapshot()

    expect(v.number().required().nonpositive().parse(1)).toMatchSnapshot()
    expect(v.number().required().nonpositive().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().nonpositive().parse(-1)).toMatchSnapshot()
    expect(v.number().required().nonpositive().parse(0)).toMatchSnapshot()

    expect(v.number().required().negative().parse(1)).toMatchSnapshot()
    expect(v.number().required().negative().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().negative().parse(-1)).toMatchSnapshot()
    expect(v.number().required().negative().parse(0)).toMatchSnapshot()

    expect(v.number().required().nonnegative().parse(1)).toMatchSnapshot()
    expect(v.number().required().nonnegative().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().nonnegative().parse(-1)).toMatchSnapshot()
    expect(v.number().required().nonnegative().parse(0)).toMatchSnapshot()

    expect(v.number().required().positiveInteger().parse(1)).toMatchSnapshot()
    expect(v.number().required().positiveInteger().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().positiveInteger().parse(-1)).toMatchSnapshot()
    expect(v.number().required().positiveInteger().parse(0)).toMatchSnapshot()

    expect(v.number().required().id().parse(1)).toMatchSnapshot()
    expect(v.number().required().id().parse(1.2)).toMatchSnapshot()
    expect(v.number().required().id().parse(-1)).toMatchSnapshot()
    expect(v.number().required().id().parse(0)).toMatchSnapshot()
  })

  test('boolean', () => {
    expect(
      v
        .boolean()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(
      v.boolean().parse(
        // @ts-expect-error
        '',
      ),
    ).toMatchSnapshot()
    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()

    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse(''),
    ).toMatchSnapshot()
    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse(0),
    ).toMatchSnapshot()
    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse(NaN),
    ).toMatchSnapshot()

    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse('2'),
    ).toMatchSnapshot()
    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse(1),
    ).toMatchSnapshot()
    expect(
      v
        .boolean()
        .required()
        // @ts-expect-error
        .parse(33.33),
    ).toMatchSnapshot()
  })

  test('date', () => {
    expect(
      v
        .date()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()
    expect(
      v.date().parse(
        // @ts-expect-error
        '',
      ),
    ).toMatchSnapshot()
    expect(
      v
        .date()
        .required()
        // @ts-expect-error
        .parse(),
    ).toMatchSnapshot()

    expect(
      v.date().required().parse(new Date('2023-07-14T03:23:28.926Z')),
    ).toMatchSnapshot()
    expect(
      v
        .date()
        .required()
        // @ts-expect-error
        .parse(new Date('2023-07-14T03:23:28.926Z').toISOString()),
    ).toMatchSnapshot()

    expect(
      v
        .date()
        .required()
        .min('2023-06-14T03:23:28.926Z', '应大于等于给定日期')
        .parse(new Date('2023-07-14T03:23:28.926Z')),
    ).toMatchSnapshot()
    expect(
      v
        .date()
        .required()
        .min(new Date('2023-09-14T03:23:28.926Z'), '应大于等于给定日期')
        .parse(new Date('2023-07-14T03:23:28.926Z')),
    ).toMatchSnapshot()

    expect(
      v
        .date()
        .required()
        .max('2023-06-14T03:23:28.926Z', '应小于等于给定日期')
        .parse(new Date('2023-07-14T03:23:28.926Z')),
    ).toMatchSnapshot()
    expect(
      v
        .date()
        .required()
        .max(new Date('2023-09-14T03:23:28.926Z'), '应小于等于给定日期')
        .parse(new Date('2023-07-14T03:23:28.926Z')),
    ).toMatchSnapshot()
  })

  test('array', () => {
    expect(
      v.array<string[]>(v.string().required()).parse(
        // @ts-expect-error
        [1, '2'],
      ),
    ).toMatchSnapshot()
    expect(
      v.array<string[]>(v.string().required()).min(3).parse(
        // @ts-expect-error
        [1, '2'],
      ),
    ).toMatchSnapshot()
    expect(
      v.array<string[]>(v.string().required()).max(2).parse(
        // @ts-expect-error
        [1, '2'],
      ),
    ).toMatchSnapshot()
  })

  test('object', () => {
    const schema = v.object<{
      id: number
      name: string
      gender: 'male' | 'female'
      images: string[]
      isAdmin?: boolean
    }>({
      id: v.number().required().id(),
      name: v.string().required().max(10),
      gender: v.string(s => s.required().enum(['male', 'female'])),
      images: v.array(a => a.element(v.string().required()).default([])),
      isAdmin: v.boolean().default(false),
    })
    expect(
      schema.parse(
        // @ts-expect-error
        { id: 1 },
      ),
    ).toMatchSnapshot()
    expect(
      schema.parse(
        // @ts-expect-error
        { id: 1, name: 'jack' },
      ),
    ).toMatchSnapshot()
    expect(
      schema.parse(
        // @ts-expect-error
        { id: 1, name: 'jack', gender: 'male' },
      ),
    ).toMatchSnapshot()
    expect(
      schema.parse({ id: 1, name: 'jack', gender: 'male', images: [''] }),
    ).toMatchSnapshot()
  })

  test('abortEarly', () => {
    const schema = v.object<{
      id: number
      images: string[]
    }>({
      id: v.number().required().id(),
      images: v.array(a =>
        a.required().element(v.string().required().url()).min(2),
      ),
    })

    expect(
      schema.parse({
        // @ts-expect-error
        images: 's',
      }),
    ).toMatchSnapshot()

    expect(
      schema.parse(
        {
          // @ts-expect-error
          images: 's',
        },
        {
          abortEarly: true,
        },
      ),
    ).toMatchSnapshot()

    expect(
      schema.parse({
        id: 1,
        images: ['http://baidu.com', 'ddd', 'dddd3'],
      }),
    ).toMatchSnapshot()

    expect(
      schema.parse(
        {
          id: 1,
          images: ['http://baidu.com', 'ddd', 'dddd3'],
        },
        {
          abortEarly: true,
        },
      ),
    ).toMatchSnapshot()
  })

  test('preserveUnknownKeys', () => {
    const schema = v.object<{
      id: number
    }>({
      id: v.number().required().id(),
    })

    expect(
      schema.parse({
        id: 1,
        // @ts-expect-error
        name: 'sss',
      }),
    ).toMatchSnapshot()

    expect(
      schema.parse(
        {
          id: 1,
          // @ts-expect-error
          name: 'sss',
        },
        {
          preserveUnknownKeys: true,
        },
      ),
    ).toMatchSnapshot()
  })

  test('case: pass1-pass2', () => {
    const schema = v
      .object<{
        password: string
        password2: string
      }>({
        password: v.string().required(),
        password2: v.string().required(),
      })
      .custom(
        ({ password, password2 }) => password === password2,
        '密码2应等于密码1',
      )

    expect(
      schema.parse({
        password: '123',
        password2: '456',
      }),
    ).toMatchSnapshot()

    expect(
      schema.parse({
        password: '123',
        password2: '123',
      }),
    ).toMatchSnapshot()

    const schema2 = v
      .object<{
        password: string
        password2: string
      }>({
        password: v.string().required(),
        password2: v.string().required(),
      })
      .custom(
        ({ password, password2 }) => password === password2,
        '密码2应等于密码1',
        'password2',
      )

    expect(
      schema2.parse({
        password: '123',
        password2: '456',
      }),
    ).toMatchSnapshot()

    expect(
      schema2.parse({
        password: '123',
        password2: '123',
      }),
    ).toMatchSnapshot()
  })

  test('optional', () => {
    const schema = v.string().required()
    expect(schema.parse('')).toMatchSnapshot()
    expect(schema.optional().parse('')).toMatchSnapshot()
    // 没有 clone 被 optional 影响了
    expect(schema.parse('')).toMatchSnapshot()
  })

  test('clone', () => {
    const schema = v.string().required()
    expect(schema.parse('')).toMatchSnapshot()
    expect(schema.clone().optional().parse('')).toMatchSnapshot()
    expect(schema.parse('')).toMatchSnapshot()
  })

  test('pickFields, omitFields', () => {
    const schema = v.object({
      id: v.number().required(),
      name: v.string().required(),
    })
    expect(
      schema.parse({
        id: 0,
        name: 'jack',
      }),
    ).toMatchSnapshot()
    expect(
      schema.clone().pickFields(['id']).parse({
        id: 0,
        // @ts-expect-error
        name: 'jack',
      }),
    ).toMatchSnapshot()
    expect(
      schema.clone().omitFields(['id']).parse({
        // @ts-expect-error
        id: 0,
        name: 'jack',
      }),
    ).toMatchSnapshot()
    expect(
      schema.clone().pickFields(['id']).parse({
        id: 0,
      }),
    ).toMatchSnapshot()
    expect(
      schema.clone().omitFields(['id']).parse({
        name: 'jack',
      }),
    ).toMatchSnapshot()
  })

  test('optionalFields, requiredFields', () => {
    const schema = v.object<{
      id: number
      name?: string
    }>({
      id: v.number().required(),
      name: v.string(),
    })
    expect(
      schema.parse({
        id: 0,
      }),
    ).toMatchSnapshot()
    expect(
      schema.clone().requiredFields(['name']).parse(
        // @ts-expect-error
        {
          id: 0,
        },
      ),
    ).toMatchSnapshot()
    expect(schema.clone().optionalFields(['id']).parse({})).toMatchSnapshot()
    expect(
      schema.clone().requiredFields().parse(
        // @ts-expect-error
        {
          id: 0,
        },
      ),
    ).toMatchSnapshot()
    expect(schema.clone().optionalFields().parse({})).toMatchSnapshot()
  })

  test('create', () => {
    const schema = v.create(_ => _.string().required())
    expect(schema.parse('1')).toMatchSnapshot()

    const schema2: VaeSchemaOf<{
      id: number
    }> = v.create(_ =>
      _.object({
        id: _.number().required(),
      }),
    )
    expect(
      schema2.parse({
        id: 1,
      }),
    ).toMatchSnapshot()
  })
})
