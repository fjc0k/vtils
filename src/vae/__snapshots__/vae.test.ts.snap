// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vae abortEarly 1`] = `
Object {
  "issues": Array [
    Object {
      "message": "id应必填",
      "path": Array [
        "id",
      ],
    },
    Object {
      "message": "images应是数组类型",
      "path": Array [
        "images",
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae abortEarly 2`] = `
Object {
  "issues": Array [
    Object {
      "message": "id应必填",
      "path": Array [
        "id",
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae abortEarly 3`] = `
Object {
  "issues": Array [
    Object {
      "message": "images.1应是一个合法的网址",
      "path": Array [
        "images",
        1,
      ],
    },
    Object {
      "message": "images.2应是一个合法的网址",
      "path": Array [
        "images",
        2,
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae abortEarly 4`] = `
Object {
  "issues": Array [
    Object {
      "message": "images.1应是一个合法的网址",
      "path": Array [
        "images",
        1,
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae array 1`] = `
Object {
  "data": Array [
    "1",
    "2",
  ],
  "success": true,
}
`;

exports[`vae array 2`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应至少包含3个元素",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae array 3`] = `
Object {
  "data": Array [
    "1",
    "2",
  ],
  "success": true,
}
`;

exports[`vae boolean 1`] = `
Object {
  "data": undefined,
  "success": true,
}
`;

exports[`vae boolean 2`] = `
Object {
  "data": false,
  "success": true,
}
`;

exports[`vae boolean 3`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应必填",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae boolean 4`] = `
Object {
  "data": false,
  "success": true,
}
`;

exports[`vae boolean 5`] = `
Object {
  "data": false,
  "success": true,
}
`;

exports[`vae boolean 6`] = `
Object {
  "data": false,
  "success": true,
}
`;

exports[`vae boolean 7`] = `
Object {
  "data": true,
  "success": true,
}
`;

exports[`vae boolean 8`] = `
Object {
  "data": true,
  "success": true,
}
`;

exports[`vae boolean 9`] = `
Object {
  "data": true,
  "success": true,
}
`;

exports[`vae case: pass1-pass2 1`] = `
Object {
  "issues": Array [
    Object {
      "message": "密码2应等于密码1",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae case: pass1-pass2 2`] = `
Object {
  "data": Object {
    "password": "123",
    "password2": "123",
  },
  "success": true,
}
`;

exports[`vae case: pass1-pass2 3`] = `
Object {
  "issues": Array [
    Object {
      "message": "密码2应等于密码1",
      "path": Array [
        "password2",
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae case: pass1-pass2 4`] = `
Object {
  "data": Object {
    "password": "123",
    "password2": "123",
  },
  "success": true,
}
`;

exports[`vae date 1`] = `
Object {
  "data": undefined,
  "success": true,
}
`;

exports[`vae date 2`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是日期类型",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae date 3`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应必填",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae date 4`] = `
Object {
  "data": 2023-07-14T03:23:28.926Z,
  "success": true,
}
`;

exports[`vae date 5`] = `
Object {
  "data": 2023-07-14T03:23:28.926Z,
  "success": true,
}
`;

exports[`vae date 6`] = `
Object {
  "data": 2023-07-14T03:23:28.926Z,
  "success": true,
}
`;

exports[`vae date 7`] = `
Object {
  "issues": Array [
    Object {
      "message": "应大于等于给定日期",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae date 8`] = `
Object {
  "issues": Array [
    Object {
      "message": "应小于等于给定日期",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae date 9`] = `
Object {
  "data": 2023-07-14T03:23:28.926Z,
  "success": true,
}
`;

exports[`vae number 1`] = `
Object {
  "data": undefined,
  "success": true,
}
`;

exports[`vae number 2`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是数值类型",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 3`] = `
Object {
  "data": 3,
  "success": true,
}
`;

exports[`vae number 4`] = `
Object {
  "data": 3,
  "success": true,
}
`;

exports[`vae number 5`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应必填",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 6`] = `
Object {
  "data": 3,
  "success": true,
}
`;

exports[`vae number 7`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是数值类型",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 8`] = `
Object {
  "data": 3,
  "success": true,
}
`;

exports[`vae number 9`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应大于或等于6",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 10`] = `
Object {
  "data": 0,
  "success": true,
}
`;

exports[`vae number 11`] = `
Object {
  "data": 0,
  "success": true,
}
`;

exports[`vae number 12`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应大于2",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 13`] = `
Object {
  "data": 0,
  "success": true,
}
`;

exports[`vae number 14`] = `
Object {
  "data": -1,
  "success": true,
}
`;

exports[`vae number 15`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 16`] = `
Object {
  "data": 1,
  "success": true,
}
`;

exports[`vae number 17`] = `
Object {
  "data": 1,
  "success": true,
}
`;

exports[`vae number 18`] = `
Object {
  "data": 1.2,
  "success": true,
}
`;

exports[`vae number 19`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 20`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 21`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个非正数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 22`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个非正数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 23`] = `
Object {
  "data": -1,
  "success": true,
}
`;

exports[`vae number 24`] = `
Object {
  "data": 0,
  "success": true,
}
`;

exports[`vae number 25`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个负数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 26`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个负数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 27`] = `
Object {
  "data": -1,
  "success": true,
}
`;

exports[`vae number 28`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个负数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 29`] = `
Object {
  "data": 1,
  "success": true,
}
`;

exports[`vae number 30`] = `
Object {
  "data": 1.2,
  "success": true,
}
`;

exports[`vae number 31`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个非负数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 32`] = `
Object {
  "data": 0,
  "success": true,
}
`;

exports[`vae number 33`] = `
Object {
  "data": 1,
  "success": true,
}
`;

exports[`vae number 34`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 35`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 36`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 37`] = `
Object {
  "data": 1,
  "success": true,
}
`;

exports[`vae number 38`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 39`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae number 40`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个正整数",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae object 1`] = `
Object {
  "issues": Array [
    Object {
      "message": "name应必填",
      "path": Array [
        "name",
      ],
    },
    Object {
      "message": "gender应必填",
      "path": Array [
        "gender",
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae object 2`] = `
Object {
  "issues": Array [
    Object {
      "message": "gender应必填",
      "path": Array [
        "gender",
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae object 3`] = `
Object {
  "data": Object {
    "gender": "male",
    "id": 1,
    "images": Array [],
    "isAdmin": false,
    "name": "jack",
  },
  "success": true,
}
`;

exports[`vae object 4`] = `
Object {
  "issues": Array [
    Object {
      "message": "images.0应必填",
      "path": Array [
        "images",
        0,
      ],
    },
  ],
  "success": false,
}
`;

exports[`vae preserveUnknownKeys 1`] = `
Object {
  "data": Object {
    "id": 1,
  },
  "success": true,
}
`;

exports[`vae preserveUnknownKeys 2`] = `
Object {
  "data": Object {
    "id": 1,
    "name": "sss",
  },
  "success": true,
}
`;

exports[`vae string 1`] = `
Object {
  "data": undefined,
  "success": true,
}
`;

exports[`vae string 2`] = `
Object {
  "data": "",
  "success": true,
}
`;

exports[`vae string 3`] = `
Object {
  "data": "hello",
  "success": true,
}
`;

exports[`vae string 4`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应必填",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 5`] = `
Object {
  "data": "hello",
  "success": true,
}
`;

exports[`vae string 6`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应必填",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 7`] = `
Object {
  "data": "",
  "success": true,
}
`;

exports[`vae string 8`] = `
Object {
  "data": "hello",
  "success": true,
}
`;

exports[`vae string 9`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应至少包含6位字符",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 10`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应最多包含2位字符",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 11`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应仅包含3位字符",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 12`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个合法的邮箱",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 13`] = `
Object {
  "data": "hello@gmail.com",
  "success": true,
}
`;

exports[`vae string 14`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个合法的网址",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 15`] = `
Object {
  "data": "https://baidu.com",
  "success": true,
}
`;

exports[`vae string 16`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应满足正则表达式/^\\\\d+$/",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 17`] = `
Object {
  "data": "234",
  "success": true,
}
`;

exports[`vae string 18`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应包含字符串dd",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 19`] = `
Object {
  "data": "23dd4",
  "success": true,
}
`;

exports[`vae string 20`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应以字符串dd开头",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 21`] = `
Object {
  "data": "dd234",
  "success": true,
}
`;

exports[`vae string 22`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应以字符串dd结尾",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 23`] = `
Object {
  "data": "234dd",
  "success": true,
}
`;

exports[`vae string 24`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个合法的手机号码",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 25`] = `
Object {
  "data": "18080000000",
  "success": true,
}
`;

exports[`vae string 26`] = `
Object {
  "issues": Array [
    Object {
      "message": ".应是一个合法的身份证号码",
      "path": Array [],
    },
  ],
  "success": false,
}
`;

exports[`vae string 27`] = `
Object {
  "data": "110101202305033210",
  "success": true,
}
`;

exports[`vae string 28`] = `
Object {
  "data": "110101202305033210",
  "success": true,
}
`;
